packages/beacon-node/src/chain/blocks/types.ts

```diff
@@ -69,75 +82,66 @@ export const getBlockInput = {
+      blockCache = this.blockInputCache.get(blockHex) ?? getEmptyBlockInputCacheEntry();
 
       blockCache.block = signedBlock;
       blockCache.blockBytes = blockBytes;
     } else {
       const {signedBlob, blobBytes} = gossipedInput;
       blockHex = toHexString(signedBlob.message.blockRoot);
-      blockCache = this.blockInputCache.get(blockHex);
-
-      // If a new entry is going to be inserted, prune out old ones
-      if (blockCache === undefined) {
-        pruneSetToMax(this.blockInputCache, MAX_GOSSIPINPUT_CACHE);
-        blockCache = {blobs: new Map<number, deneb.BlobSidecar>(), blobsBytes: new Map<number, Uint8Array | null>()};
-      }
+      blockCache = this.blockInputCache.get(blockHex) ?? getEmptyBlockInputCacheEntry();
 
       // TODO: freetheblobs check if its the same blob or a duplicate and throw/take actions
-      blockCache.blobs.set(signedBlob.message.index, signedBlob.message);
-      // easily splice out the unsigned message as blob is a fixed length type
-      blockCache.blobsBytes.set(signedBlob.message.index, blobBytes?.slice(0, BLOBSIDECAR_FIXED_SIZE) ?? null);
+      blockCache.blobsCache.set(signedBlob.message.index, {
+        blobSidecar: signedBlob.message,
+        // easily splice out the unsigned message as blob is a fixed length type
+        blobBytes: blobBytes?.slice(0, BLOBSIDECAR_FIXED_SIZE) ?? null,
+      });
     }
 
     this.blockInputCache.set(blockHex, blockCache);
-    const {block: signedBlock, blockBytes} = blockCache;
+    const {block: signedBlock, blockBytes, blobsCache, availabilityPromise, resolveAvailability} = blockCache;
 
     if (signedBlock !== undefined) {
       // block is available, check if all blobs have shown up
       const {slot, body} = signedBlock.message;
       const {blobKzgCommitments} = body as deneb.BeaconBlockBody;
       const blockInfo = `blockHex=${blockHex}, slot=${slot}`;
 
-      if (blobKzgCommitments.length < blockCache.blobs.size) {
+      if (blobKzgCommitments.length < blobsCache.size) {
         throw Error(
-          `Received more blobs=${blockCache.blobs.size} than commitments=${blobKzgCommitments.length} for ${blockInfo}`
+          `Received more blobs=${blobsCache.size} than commitments=${blobKzgCommitments.length} for ${blockInfo}`
         );
       }
-      if (blobKzgCommitments.length === blockCache.blobs.size) {
-        const blobSidecars = [];
-        const blobsBytes = [];
-
-        for (let index = 0; index < blobKzgCommitments.length; index++) {
-          const blobSidecar = blockCache.blobs.get(index);
-          if (blobSidecar === undefined) {
-            throw Error(`Missing blobSidecar at index=${index} for ${blockInfo}`);
-          }
-          blobSidecars.push(blobSidecar);
-          blobsBytes.push(blockCache.blobsBytes.get(index) ?? null);
-        }
 
+      if (blobKzgCommitments.length === blobsCache.size) {
+        const allBlobs = getBlockInputBlobs(blobsCache);
+        resolveAvailability(allBlobs);
```

Comment:
The scope of this function is growing too much, not a fan. See comment on https://github.com/ChainSafe/lodestar/pull/6087/files#diff-371b6c76d2e10d7638874dbf97610d25926b5f937e009d8bf5a3aa2f8f17d6b7R63